# **아파트 실거래가 예측 프로젝트**


## **Stage 2. 문제 및 모델링의 이해 - 아파트 실거래가를 예측하기 위해 필요한 것**
<img src='http://drive.google.com/uc?export=view&id=1KpVIVt81WBwyz5bAAyvm0Lszl09ZTMjj' /><br>


### **INTRO**
지난 스테이지 **Stage 1. 문제 기반 실무 학습(PBL) 개요 - 아파트 실거래가 예측 프로젝트** 에서 **프로젝트를 파악**하고, **판다스 라이브러리를 사용**하여 데이터를 불러와서 제출했습니다 😀  

아파트 실거래가를 예측하기 위해 **가장 먼저 필요한 것**은 무엇일까요?    
**모델 학습을 위한 train 데이터**와 **모델 평가를 위한 test 데이터**가 필요합니다 😎

이번 스테이지는 변수들의 **기술적 통계**을 확인하고, **train 및 test 데이터셋을 구성**하여 **선형회귀 모델을 적용**할 예정입니다.    

이제 판다스 라이브러리를 임포트하고, 데이터를 불러와봅시다~


In [1]:
import pandas as pd

In [2]:
train = pd.read_csv('/content/drive/MyDrive/데이크루4기_먼스크루/data/dacon/train.csv') 
test = pd.read_csv('/content/drive/MyDrive/데이크루4기_먼스크루/data/dacon/test.csv')
park = pd.read_csv('/content/drive/MyDrive/데이크루4기_먼스크루/data/dacon/park.csv')
day_care_center = pd.read_csv('/content/drive/MyDrive/데이크루4기_먼스크루/data/dacon/day_care_center.csv')
submission = pd.read_csv('/content/drive/MyDrive/데이크루4기_먼스크루/data/dacon/submission.csv')

Mounted at /content/drive


### **1. 변수 조사 및 기술적 통계**

우선 주어진 데이터를 먼저 분석해볼게요!    
head() 함수로 데이터 구성을 다시 확인해볼까요?

#### 1-1. 변수 조사

In [3]:
train.head()

transaction_id  apartment_id   city dong   jibun       apt  \
0               0          7622  서울특별시  신교동    6-13  신현(101동)   
1               1          5399  서울특별시  필운동     142    사직파크맨션   
2               2          3578  서울특별시  필운동   174-1    두레엘리시안   
3               3         10957  서울특별시  내수동      95     파크팰리스   
4               4         10639  서울특별시  내수동  110-15      킹스매너   

             addr_kr  exclusive_use_area  year_of_completion  \
0  신교동 6-13 신현(101동)               84.82                2002   
1     필운동 142 사직파크맨션               99.17                1973   
2   필운동 174-1 두레엘리시안               84.74                2007   
3       내수동 95 파크팰리스              146.39                2003   
4    내수동 110-15 킹스매너              194.43                2004   

   transaction_year_month transaction_date  floor  transaction_real_price  
0                  200801            21~31      2                   37500  
1                  200801             1~10      6                   20000  
2                  200801             1~10      6                   38500  
3                  200801            11~20     15                  118000  
4                  200801            21~31      3                  120000

13개의 컬럼으로 이루어진 데이터네요!

#### 1-2. 기술 통계
판다스의 describe() 함수는 숫자형(int, float 등) 데이터에 대한 기술 통계를 제공합니다.    
describe()로 숫자형 데이터에 대한 기술 통계를 확인해볼까요?


##### **Inst.**
그럼 먼저 앞으로 우리가 분석해볼 데이터를 대략적으로 확인해볼까요?

1. train 데이터를 describe()를 사용하여 요약하세요.


In [ ]:
# 1. train 데이터를 describe()를 사용하여 요약하세요.


##### **Hint.**
- <font color="orange">describe()</font> 로 패키지를 사용할 수 있습니다. 
    - <font color="orange">OOOOO.OOOOOOO()</font>


##### **Solution.**


In [4]:
# 1. train 데이터를 describe()를 사용하여 요약하세요.
train.describe()

transaction_id  apartment_id  exclusive_use_area  year_of_completion  \
count    1.216553e+06  1.216553e+06        1.216553e+06        1.216553e+06   
mean     6.091530e+05  6.299685e+03        7.816549e+01        1.998297e+03   
std      3.526198e+05  3.581169e+03        2.915113e+01        8.941347e+00   
min      0.000000e+00  0.000000e+00        9.260000e+00        1.961000e+03   
25%      3.041380e+05  3.345000e+03        5.976000e+01        1.993000e+03   
50%      6.082760e+05  5.964000e+03        8.241000e+01        1.999000e+03   
75%      9.124140e+05  9.436000e+03        8.497000e+01        2.005000e+03   
max      1.234827e+06  1.265800e+04        4.243200e+02        2.017000e+03   

       transaction_year_month         floor  transaction_real_price  
count            1.216553e+06  1.216553e+06            1.216553e+06  
mean             2.012984e+05  9.343291e+00            3.822769e+04  
std              2.905732e+02  6.606500e+00            3.104898e+04  
min              2.008010e+05 -4.000000e+00            1.000000e+02  
25%              2.010100e+05  4.000000e+00            1.900000e+04  
50%              2.013120e+05  8.000000e+00            3.090000e+04  
75%              2.015110e+05  1.300000e+01            4.700000e+04  
max              2.017110e+05  8.000000e+01            8.200000e+05

주어진 데이터를 분석하는 방법은 크게 3가지 정도 있습니다.
원하는 정보를 얻기 위해서 각 코드의 특징을 알아두면 좋습니다 :)     
describe를 이용하면 값의 범위를 확인할 수 있습니다.      
예를들어 exclusive_use_area을 확인하면 값의 범위가 최소 9.26에서 최대 424.32인걸 알 수 있습니다.     

단순 ID와 날짜를 제외하고 해당 데이터의 exclusive_use_area, year_of_completion, floor, transaction_real_price을 중점적으로 보면 좋겠네요!

### **2. 모델링을 위한 데이터셋 준비(Train/Test)**

추가적인 데이터를 사용하기 위해 서울특별시에서 거래된 데이터만을 추출하여 모델 train에 이용합니다.

In [8]:
train = train[train['city'] == '서울특별시']
test = test[test['city'] == '서울특별시']

앞서 설명한 중요 column만 분석에 이용하기 위해 불필요한 column은 제외합니다.

In [9]:
drop_columns =['transaction_id', 'apartment_id', 'city', 'dong', 'jibun', 'apt','addr_kr','transaction_date','transaction_year_month']
train.drop(drop_columns, axis=1,inplace=True)
test.drop(drop_columns, axis=1,inplace=True)

원하는 컬럼이 제거됐는지 확인합니다.

In [10]:
train.head()

exclusive_use_area  year_of_completion  floor  transaction_real_price
0               84.82                2002      2                   37500
1               99.17                1973      6                   20000
2               84.74                2007      6                   38500
3              146.39                2003     15                  118000
4              194.43                2004      3                  120000

우리는 이제 실거래가를 예측하기 위한 사전 준비는 끝났습니다.     
따라서 실거래가를 나머지 변수들로 값을 예측하는 모델을 구성해야합니다.
   
이에 맞춰 아래와 같이 독립변수로 이루어진 X와 예측값인 종속변수로 이루어진 y를 선언합니다.

In [11]:
X = train.drop('transaction_real_price', axis=1)
y = train[['transaction_real_price']]

   
일반적으로 모델을 학습할 때, 
- train: 모델의 학습을 위한 데이터
- validation: 모델의 파라미터 업데이트를 위한 데이터
- test: 모델의 평가를 위한 데이터(학습에 사용하지 않음)    

로 나누어 진행합니다.
아래 그림을 참고하시면, 이해하기 좋습니다.
![train_test](https://cdn.imweb.me/upload/S202105076e626e4618b27/0f0c2df8f341c.png)

데이터를 나눠서 사용하면 모델의 학습에서 활용되지 않은 test 데이터로 타당하게 모델을 평가할 수 있습니다.  
sklearn의 train_test_split 라이브러리를 임포트하여 사용하겠습니다.  



In [12]:
from sklearn.model_selection import train_test_split

##### **Inst.**
이번에서 train과 test 데이터로만 데이터를 나눠보겠습니다.  <br/> train_test_split() 함수를 이용하여 데이터를 나눠볼까요?  🤗
1. train_test_split()을 이용하여 데이터를 분리합니다. (단, test_size는 0.2, random_state는 42로 설정합니다)

In [13]:
# 1. train_test_split()을 이용하여 데이터를 분리합니다. (단, test_size는 0.2, random_state는 42로 설정합니다)


##### **Hint.**
- train_test_split 함수의 input에는 **x, y, test_size, random_state**를 설정해야 합니다.


##### **Solution.**

In [14]:
# 1. train_test_split()을 이용하여 데이터를 분리합니다. (단, test_size는 0.2, random_state는 42로 설정합니다)
train_X, test_X, train_Y, test_Y = train_test_split( X, y, test_size=0.2, random_state=42)

8:2의 비율로 잘 나눠졌는지 확인합니다.    
shape 함수로 데이터의 행과 열을 확인할 수 있어요!
- test_size는 테스트 데이터셋의 비율을 의미해요. (default=0.25)
- random_state는 데이터 분할시 셔플이 이루어지는데 이를 위한 시드값을 의미해요. 실제로 같은 시드값을 사용하면 같은 data set을 사용할 수 있어요.

In [15]:
print(train_X.shape, test_X.shape, train_Y.shape, test_Y.shape)

(593828, 3) (148457, 3) (593828, 1) (148457, 1)


### **3. 기초 모델링: 선형 회귀**

우선적으로 학습할 모델은 선형회귀입니다.    
선형회귀는 
$$Y=WX+B$$
와 같이 독립변수 X에 가중치 파라미터를 곱하고, 상수 바이어스를 더해서 종속변수 Y를 예측할 수 있어요!     
sklearn 라이브러리의 LinearRegression을 사용하기 위해 먼저 임포트할게요.

In [16]:
from sklearn.linear_model import LinearRegression

    
##### **Inst.**
LinearRegression() 함수로 선형회귀 모델을 생성하고, fit() 함수로 모델을 학습할 수 있어요.    
선형회귀 모델을 한 번 학습시켜볼까요? 🤗

1. lrmodel 변수에 선형회귀 모델을 선언합니다.
2. lrmodel에 train_X, train_Y를 fit 합니다.

In [ ]:
# 1. lrmodel 변수에 선형회귀 모델을 선언합니다.


In [ ]:
# 2. lrmodel에 train_X, train_Y를 fit 합니다.


##### **Hint.**
- 선형회귀 모델은 LinearRegression()함수를 이용해요.
- 선형회귀 모델을 fit 하기 위해서는 x,y 데이터가 필요해요

##### **Solution.**



In [17]:
# 1. lrmodel 변수에 선형회귀 모델을 선언합니다.
lrmodel = LinearRegression()

In [18]:
# 2. lrmodel에 train_X, train_Y를 fit 합니다.
lrmodel.fit(train_X, train_Y)

LinearRegression()

학습을 완료하고 나서 실제 우리의 모델이 얼마나 완벽한지 확인하기 위해 점수를 계산합니다.    
프로젝트에서는 RMSE(
평균 제곱근 오차)가 평가 기준으로 사용하고 있어요.   
RMSE 식은 다음과 같이 MSE(평균 제곱 오차)에 제곱근을 취한 형태입니다! 
$${\displaystyle \operatorname {RMSE} ({\hat {\theta }})={\sqrt {\operatorname {MSE} ({\hat {\theta }})}}={\sqrt {\operatorname {E} (({\hat {\theta }}-\theta )^{2})}}.}$$

RMSE를 계산하기 위한 mean_squared_error 라이브러리를 먼저 임포트해야 해요!

In [ ]:
# RSME 성능평가 계산을 위한 라이브러리
from sklearn.metrics import mean_squared_error

mean_squared_error로 RMSE를 계산하는 함수는 MSE에 제곱근을 취하여 아래와 같이 작성될 수 있어요!

In [ ]:
def RMSE(y, y_pred):
    rmse = mean_squared_error(y, y_pred) ** 0.5
    return rmse

이제 RMSE 함수로 모델을 예측해봅시다!

In [ ]:
print(RMSE(lrmodel.predict(train_X), train_Y))
print(RMSE(lrmodel.predict(test_X), test_Y))

24845.657717237664
24762.40551060151


RMSE 점수는 24845 네요!

조금 아쉬운 점수인가요?   
그렇다면 더 높은 정확도를 가진 모델을 찾기 위해서 stage3에서 이어서 진행보도록 하겠습니다.    
<img src='http://drive.google.com/uc?export=view&id=161p7_9jSKgi7s305nEmvwriJL_c0Vkmo' /><br>

여기까지 진행하시느라 고생하셨습니다 😀